lib.path.hasPrefix, lib.path.removePrefix: init

infinisil opened the pull request:
###### Description of changes

Adds two new path library functions:
- `lib.path.hasPrefix :: Path -> Path -> Bool`: Whether the second path is a prefix of the first path, or equal to it:
  ```nix
  lib.path.hasPrefix ./. ./lib
  -> true
  ```
- `lib.path.removePrefix :: Path -> Path -> String`: Remove a first prefix path from a second path, the result is a normalised subpath string, see [`lib.path.subpath.normalise`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.path.subpath.normalise):
  ```nix
  lib.path.removePrefix ./. ./lib
  -> "./lib"
  ```

This relates to other work in [the path library effort](https://github.com/NixOS/nixpkgs/issues/210426).
This PR depended on https://github.com/NixOS/nixpkgs/pull/221204, which is now merged.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles: 

###### Things done
- [x] Wrote docs
- [x] Wrote tests

infinisil commented on lib/path/default.nix:
```diff
@@ -92,8 +95,114 @@ let
 
+  # Takes a Nix path value and deconstructs it into:
+  # - root: The filesystem root of the path, generally `/`
+  # - components: All the path components of the path
```
Well, [POSIX calls it path components](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_272), [RFC 3986 (URI) calls it segments](https://www.rfc-editor.org/rfc/rfc3986#section-3.3), though I don't think that's the same as unix paths. As for libraries:
- Python: [segments](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath)
- Nodejs: [segments](https://nodejs.org/api/path.html#pathresolvepaths)
- Rust: [components](https://doc.rust-lang.org/std/path/struct.Path.html#method.components)
- C++: weakly [components](https://en.cppreference.com/w/cpp/filesystem/path/remove_filename)
- Java: [name element](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#iterator--)
- Objective-C/Swift: [components](https://developer.apple.com/documentation/foundation/nsstring/1414489-pathcomponents)
- R: [components](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/normalizePath)

It looks like "segments" is more Web and URL related in general, whereas "components" is more focused on local paths. I'd say "components" should be fine here

fricklerhandwerk replied:
Alright, the word itself has some degree of freedom within reasonable proximity to the intended meaning. It could as well be part, item, node, piece, whatever. It should stay close to what's in common use though, therefore component and segment are valid candidates.

I don't think use of a term in context of Unix paths vs. URIs is a meaningful distinction, they describe the same concept from slightly different angles, and therefore in consider it a weak argument.

A stronger argument would be about what the word originally means, and if two words fit the meaning equally well, which one is easier to write and pronounce.

And then, indeed since [segment](https://en.m.wiktionary.org/wiki/segment) *also* refers to a "piece cut out or off" (which could be any subsequence of the items comprising a path) it's less suitable than [component](https://en.m.wiktionary.org/wiki/component), which refers to the things that are "put together" to make the whole.

infinisil commented on lib/path/default.nix:
```diff
@@ -92,8 +95,114 @@ let
 in /* No rec! Add dependencies on this file at the top. */ {
 
+  /*
+  Whether the second path is a descendant of the first path, or equal to it.
```
Paths do represent trees though, I don't think anybody would be confused by that. In contrast, calling these functions `isPrefixOf` or `hasPrefix` has a good chance of confusing users with the string-based `lib.strings.hasPrefix`, `lib.strings.hasSuffix`, etc.

fricklerhandwerk replied:
> Paths do represent trees though

I'm genuinely confused now. How is a path not a sequence but a tree?

And while common parlance conflates the two, that kind of fuzzy thinking leads to fuzzy thinking that we really want to avoid.

infinisil replied:
I'm not saying a path _is_ a tree, I'm saying it represents one. E.g. if you import /foo into the store is will also import /foo/bar because it's a descendant.

But I now see that this doesn't really hold when the paths don't exist, a parent-descendant relationship doesn't make sense then. So yes agreed this naming isn't ideal.

fricklerhandwerk replied:
Okay, you did not say it *is* one in terms of equivalence, but it also doesn't represent one. A path *leads* to a file system object, which *can* be a tree – or a single file. That is something very different than representation. A string of some shape may represent a path, or a drawing of some sort may represent a tree.

infinisil replied:
We're getting down to definitions now, arguably the source of most arguments. Here's the definitions for "represent" in the Merriam-Webster dictionary: https://www.merriam-webster.com/dictionary/representing

> 2.  to serve as a sign or symbol of 
>     > the flag represents our country

This is the definition I was using.

But you're right, in programming we often use the term "represent" to mean different representations of the same thing, you can often convert from one representation to another and back without information loss. I'm not sure which dictionary definition fits that though.

infinisil replied:
We could use `startsWith /foo /foo/bar` insteaed, which mirrors Rust's [`starts_with`](https://doc.rust-lang.org/std/path/struct.Path.html#method.starts_with).

Only then it's a bit unclear what the `isProperDescendantOf` version should be. Maybe we can just leave that function out, users can always have an additional `!=` check when they need that version.

Another problem then would be what [`lib.path.difference`](https://github.com/NixOS/nixpkgs/pull/209375) should return, `commonStart` sounds odd, maybe we could just use `prefix` then at the cost of some consistency.

---

I guess calling this `hasPrefix` might indeed be the best option locally, but I really don't like the potential confusion with `lib.strings.hasPrefix`, which notably has totally unexpected behavior for paths:

```
nix-repl> lib.strings.hasPrefix ./lib ./lib/cli.nix
false
```

And it's exported from `lib`:
```
nix-repl> lib.hasPrefix ./lib ./lib/cli.nix
false
```

We could perhaps change `hasPrefix` to give an error for path values, but that's a breaking change. Though I think that might be the best way forward because honestly most `lib.strings` functions don't make a lot of sense to call on paths directly, they all implicitly convert them to store paths.

roberth replied:
> But I now see that this doesn't really hold when the paths don't exist

Paths are always fictional things for which existence is irrelevant. It's the file system object it references that matters, and I have yet to come across a standard library that has a type of path that guarantees to be backed by a real file system object. Those objects are files or handles, and usually these objects forget the path. I don't see a problem with terminology that operates on a fictional tree of all paths.

I think we're all clever enough to find arguments for and against everything or weasel our way out. For example: If fictional isn't good, we can't have `hasPrefix` either, because it operates on a fictional set of possible prefixes that may have been used to construct the path.

I guess the real question is whether we can pick an abstraction that works reasonably well for all the operations we want to have on paths?

fricklerhandwerk replied:
@roberth I don't think anyone here is trying to weasel their way into being right, and instead we try to find the most convincing answer. 🙂

> I really don't like the potential confusion with `lib.strings.hasPrefix`

@infinisil We brought this up multiple times and while I see your concern, I consider it lower priority. The attribute-set-namespace explicitly says one thing is for strings, the other for paths. Relative paths are not regular strings. We should plaster that all over documentation. Yes, one can discard the namespace prefix on import, but you'd have to do it deliberately in each file if you follow best practices.

> We could perhaps change `[string.]hasPrefix` to give an error for path values, but that's a breaking change.

> most `lib.strings` functions don't make a lot of sense to call on paths directly, they all implicitly convert them to store paths.

Fully agree.

> Paths are always fictional things for which existence is irrelevant.

Exactly, although with existence we may mean two different things:
- existence of a path in the directory tree, in the graph-theoretic sense
- existence of a path in the sense of being written down in some representation

The former indeed only starts to matter once we want to traverse the path in the directory tree.
What we care about in this library is only the latter, and suddenly the conceptual model matters, and that is – sorry for being repetitive – a sequence of path components.

infinisil replied:
> Yes, one can discard the namespace prefix on import, but you'd have to do it deliberately in each file if you follow best practices.

The thing is, people aren't following best practices! In the grand scheme of things, what we do here is highly irregular, no third party takes that much care writing their code.

In particular, the (arguably anti-)pattern of using `with lib;` is all over the place. If both `lib.path.hasPrefix` and `lib.hasPrefix` exist, better make sure to never forget the `path.` prefix, otherwise you'll get wrong results like the one I showed. And this is even more likely to happen because traditionally everything has been re-exported from `lib.*`, meaning people may assume that `lib.path.hasPrefix` also follows that, and sure enough `lib.hasPrefix` exists.

Because of that I don't think we can "just" have `lib.path.hasPrefix` be a thing. The only reasonably thing that I can see working is to deprecate support for passing paths to `lib.strings.hasPrefix`.

Additionally we could also deprecate `lib.hasPrefix` and require people to use `lib.strings.hasPrefix` instead, but this would break _a lot_ of code so I'd rather not unless/until we're sure that these `lib.*` are truly a bad idea.

infinisil replied:
I now opened https://github.com/NixOS/nixpkgs/pull/221204 to implement my suggestion of deprecating `hasPrefix` for paths.

infinisil replied:
I now switched this PR to use `hasPrefix` and `hasProperPrefix`, which I think is okay if the above PR is merged

roberth replied:
Weak typing was particularly bad for `hasPrefix`:

```
nix-repl> lib.hasPrefix ~/. ~/h/nixpkgs
^C^C^C^C^C^C^C^C^C^C^Z
[1]+  Stopped                 nix repl

$ kill -9 %%
```

Friends don't let friends add their home directory to the store ;)

infinisil commented on lib/path/default.nix:
```diff
@@ -92,8 +95,114 @@ let
+
+        isDescendantOf p q <-> exists s . p + ("/" + s) == q
+
+  - isDescendantOf is a non-strict partial order over the set of all path values:
+
+    https://en.wikipedia.org/wiki/Partially_ordered_set#Non-strict_partial_order
```
Gonna change this a bit, markdown isn't rendered here

infinisil commented on lib/path/default.nix:
```diff
@@ -92,8 +95,114 @@ let
+    isDescendantOf /. /foo
+    => true
+  */
+  isDescendantOf = ancestor: descendant:
```
I do like how the names make the ordering clear. I'll change this to `ancestor` and `potentialDescendant` though

infinisil commented on lib/path/default.nix:
```diff
@@ -92,8 +95,114 @@ let
+      "lib.path.isDescendantOf: Second argument is of type ${typeOf ancestor}, but a path value was expected";
+    assert assertMsg
+      (ancestor'.root == descendant'.root)
+      "lib.path.isDescendantOf: First argument \"${toString ancestor}\" (root ${toString ancestor'.root}) has a different filesystem root than the second argument \"${toString descendant}\" (root ${toString descendant'.root})";
```
If you mean regarding newlines: https://github.com/NixOS/nixpkgs/pull/208887#discussion_r1069737602

infinisil commented:
> This looks ready to me now.

This PR depends on https://github.com/NixOS/nixpkgs/pull/221204 though, see https://github.com/NixOS/nixpkgs/pull/210423#discussion_r1136097155

roberth reviewed (approved):
I've read and ponder for a long time now.
Some thoughts about making the error scenario about unequal roots handle-able, but these functions _are_ the right tool for the job, because most of the time, equality of roots is an ambient property that can be assumed and forgotten by the caller exactly because it's checked here.
We might want to help with detecting unequal roots ahead of time for functions that need to care about that for whatever reason, but I don't think we have an example of that yet, and it doesn't have to be in scope for this pr.

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -100,6 +102,20 @@ let
 
+  # Deconstruct a path value type into:
+  # - root: The filesystem root of the path, generally `/`
+  # - components: All the path components of the path
```
```suggestion
  # - components: All the components of the path
```

```suggestion
  # - components: All the path's components
```

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -149,6 +165,105 @@ in /* No rec! Add dependencies on this file at the top. */ {
+    prefix:
+    # The path that that might start with the given prefix
+    path:
+    let
```
Does it make sense to extract that into a helper that does the type validation and fancy error messages and just produces the results required to work with? Then the two implementations would be much closer to what their invariants say about them, especially they'd clearly look similar in the exact way they are.

infinisil replied:
Seems to work!

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -100,6 +102,20 @@ let
+  # because it can distinguish different filesystem roots
+  deconstructPath =
+    let
+      go = components: path:
```
The naming nit from the other PR still holds 😛

infinisil replied:
I'll change it to `recurse` [as in the other PR](https://github.com/NixOS/nixpkgs/pull/209375#discussion_r1069220818) :)

infinisil commented on lib/path/default.nix:
```diff
@@ -149,6 +209,100 @@ in /* No rec! Add dependencies on this file at the top. */ {
+    removePrefix /. /foo
+    => "./foo"
+  */
+  removePrefix = withTwoDeconstructedPaths "lib.path.removePrefix" (prefix: deconPrefix: path: deconPath:
```
I now added a third function to this PR, `removePrefix` because it's very related to the other two and can share a lot of the code. Previously in https://github.com/NixOS/nixpkgs/pull/200718 I called it `relativeTo`

infinisil committed:
lib.path.append: Add a law

With removePrefix introduced in a future commit this law can then be
used to derive

        removePrefix p (append p s) == subpath.normalise s
        => (wrap with append)
        append p (removePrefix p (append p s)) == append p (subpath.normalise s)
        => (append is not influenced by subpath normalisation)
        append p (removePrefix p (append p s)) == append p s
        => (substitute q = append p s)
        append p (removePrefix p q) == q

Not included in the docs because it's not that important, just shows
that the first statement is more general than the second one (because
this derivation doesn't work the other way)

infinisil committed:
lib.path.hasPrefix: init

infinisil committed:
lib.path.removePrefix: init

fricklerhandwerk commented on lib/path/default.nix:
```diff
@@ -100,6 +103,54 @@ let
+        else recurse ([ (baseNameOf base) ] ++ components) (dirOf base);
+    in recurse [] path;
+
+  # Type: String -> (DeconstructedPath -> DeconstructedPath -> a) -> (Path -> (Path -> a))
+  #
+  # Turn a binary path function on deconstructed paths into a function on
+  # non-deconstructed paths. The resulting function takes one path after
+  # another, deconstructing it, and caching it before taking the next path as
+  # an argument, allowing the deconstruction to be reused between calls with
+  # the same first argument. With both paths passed, a check ensures that they
+  # have the same filesystem root, before the passed function on deconstructed
+  # paths is called. The first argument is the context for error messages.
+  withTwoDeconstructedPaths = context: f:
+    path1:
+    assert assertMsg
+      (isPath path1)
+      "${context}: First argument is of type ${typeOf path1}, but a path was expected";
+    let
+      deconPath1 = deconstructPath path1;
+    in
+      path2:
+      assert assertMsg
+        (isPath path2)
+        "${context}: Second argument is of type ${typeOf path2}, but a path was expected";
+      let
+        deconPath2 = deconstructPath path2;
+      in
+        assert assertMsg
+        (deconPath1.root == deconPath2.root) ''
+          ${context}: Filesystem roots must be the same for both paths, but paths with different roots were given:
+              first argument: "${toString path1}" (root "${toString deconPath1.root}")
+              second argument: "${toString path2}" (root "${toString deconPath2.root}")'';
+        f deconPath1 deconPath2;
```
This is way too clever to be maintainable.

Every day it looks more like what you actually want to implement is error tracing and type checking at the Nix language level.

Apparently what we really need here is
1. generic type assertions, this seems reasonable to extract into a helper function
2. the root check, but there is no reason to hard-code it to be about pairs of paths

   actually it's specific enough to duplicate it across the two functions, as it was originally.

roberth replied:
I agree.

> generic type assertions

This would perhaps be nice to have in the language, although we'd have to choose between having a static type system or slow evaluation or more infinite recursions. I won't go into more detail here.

To move this forward, I suggest removing this function, which is a non-abstraction, and use more general checks at the call sites.

roberth commented on lib/path/default.nix:
```diff
@@ -149,6 +174,104 @@ in /* No rec! Add dependencies on this file at the top. */ {
+    => true
+  */
+  hasPrefix =
+    path1:
+    assert assertMsg
+      (isPath path1)
+      "lib.path.hasPrefix: First argument is of type ${typeOf path1}, but a path was expected";
+    let
+      path1Deconstructed = deconstructPath path1;
+    in
+      path2:
+      assert assertMsg
+        (isPath path2)
+        "lib.path.hasPrefix: Second argument is of type ${typeOf path2}, but a path was expected";
+      let
+        path2Deconstructed = deconstructPath path2;
+      in
+        assert assertMsg
+        (path1Deconstructed.root == path2Deconstructed.root) ''
+          lib.path.hasPrefix: Filesystem roots must be the same for both paths, but paths with different roots were given:
+              first argument: "${toString path1}" (root "${toString path1Deconstructed.root}")
+              second argument: "${toString path2}" (root "${toString path2Deconstructed.root}")'';
```
```suggestion
    path1: assert assertMsg (isPath path1) "lib.path.hasPrefix: First argument is of type ${typeOf path1}, but a path was expected";
    path2: assert assertMsg (isPath path2) "lib.path.hasPrefix: Second argument is of type ${typeOf path2}, but a path was expected";
      let
        path1Deconstructed = deconstructPath path1;
        path2Deconstructed = deconstructPath path2;
      in
        assert assertMsg (path1Deconstructed.root == path2Deconstructed.root) ''
          lib.path.hasPrefix: Filesystem roots must be the same for both paths, but paths with different roots were given:
              first argument: "${toString path1}" with root "${toString path1Deconstructed.root}"
              second argument: "${toString path2}" with root "${toString path2Deconstructed.root}"'';

```

Changes:
 - a slight refactor that saves an Env and is a bit more readable in my opinion.
 - removed parentheses from the error message.
 - added a newline before the actual work starts for readability.

Consider these for `removePrefix` as well.

roberth replied:
Alright then I guess. ViewPatterns would be nice.

I fell into this rabbit hole by the way, while unaware of the uncached transformation https://github.com/NixOS/nix/issues/8211

It stops just short of ViewPatterns, unless `@=` isn't deduplicated or has a non-deduped variant (though I guess at that point we need keywords instead of operator soup).

infinisil replied:
Subscribed to that issue, though I don't want to get into that rabbit hole myself right now :)

I applied your other suggestion to remove the parenthesis from the error message though

roberth commented on lib/path/default.nix:
```diff
@@ -149,6 +174,104 @@ in /* No rec! Add dependencies on this file at the top. */ {
+
+  Laws:
+
+  - Inverses `append` for normalised subpaths:
```
```suggestion
  - Inverts `append` for normalised subpaths:
```

Inverses is the plural of the noun; not a verb.

infinisil replied:
Applied

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-47/27387/1

infinisil commented on lib/path/default.nix:
```diff
@@ -149,6 +174,104 @@ in /* No rec! Add dependencies on this file at the top. */ {
+    removePrefix /. /foo
+    => "./foo"
+  */
+  removePrefix =
```
I'd like to just rethink this a little bit. While working on https://github.com/NixOS/nixpkgs/pull/222981 I realized the frequent need to operate over the components of a subpath. This function as implemented here however returns the subpath as a string, which would then have to be parsed into a list of components again, which is a bit wasteful.

So I'm tilting towards an API like this:
- `lib.path.components.removePrefix /foo /foo/bar -> [ "bar" ]`
- `lib.path.components.toSubpath [ "bar" ] -> "./bar"`
- `lib.path.components.fromSubpath "./bar" -> [ "bar" ]`
- `lib.path.parts.deconstruct /foo -> { root = /; components = [ "foo" ]; }`

Needs more thought, so I don't consider this PR ready to merge anymore

fricklerhandwerk replied:
That API does make sense though.

infinisil commented:
I now opened https://github.com/NixOS/nixpkgs/pull/237610 only containing `hasPrefix`, so that can be merged before figuring out things here

infinisil commented:
I opened https://github.com/NixOS/nixpkgs/pull/238013 for a new `removePrefix` design. I'll close this PR in favor of the above two PR's.
